src/core/Song.cpp

```diff
@@ -415,6 +409,68 @@ void Song::processNextBuffer()
 	}
 }
 
+void Song::processAutomations(const TrackList &tracklist, MidiTime timeStart, fpp_t frames, int tcoNum)
+{
+	QVector<AutomationTrack*> tracks;
+
+	if(m_playMode == Mode_PlaySong)
+	{
+		tracks << m_globalAutomationTrack;
+	}
+	for( Track* track : tracklist)
+	{
+		if (track->type() == Track::AutomationTrack || track->type() == Track::HiddenAutomationTrack)
+		{
+			tracks << dynamic_cast<AutomationTrack*>(track);
+		}
+	}
+
+	MidiTime time_end = timeStart + static_cast<int>(frames / Engine::framesPerTick());
+
+	QVector<TrackContentObject*> tcos;
+	for (Track* track: tracks)
+	{
+		if(track->isMuted())
+		{
+			continue;
+		}
+
+		int tco_begin = 0, tco_end = track->numOfTCOs();
+		if( tcoNum >= 0 )
+		{
+			tco_begin = tcoNum;
+			tco_end = tcoNum + 1;
+		}
+
+		for (int i = tco_begin; i < tco_end; i++)
+		{
+			TrackContentObject* tco = track->getTCO(i);
+
+			// Skip this TCO if it's muted or not contained in the time period we're processing
+			if( tco == NULL || tco->isMuted() || tco->startPosition() > time_end)
```

Comment:
I agree, but I didn't write that part, I just moved it and didn't dare remove check introduced by `@tobydox` 9 years ago. Feel free to add something like a `qCritical` message or to just remove the check.